doc/source/scripts/gen_topologyparser_attrs.py

```diff
@@ -64,101 +65,170 @@
+            return singular
 
+        def _atomgroup(name: str, singular: str, description: str) -> str:
+            return name
 
-class TopologyAttrs(TableWriter):
-    headings = ("Atom", "AtomGroup", "Description", "Supported formats")
-    filename = "generated/topology/topologyattrs.txt"
+        def _description(name: str, singular: str, description: str) -> str:
+            return description
 
-    def __init__(self, attrs):
-        self.attrs = attrs
-        super(TopologyAttrs, self).__init__()
+        def _supported_formats(name: str, singular: str, description: str) -> str:
+            return ", ".join(sorted(attrs[name]))
 
-    def _set_up_input(self):
-        return sorted(
-            [x, *y] for x, y in NON_CORE_ATTRS.items() if x not in MANDATORY_ATTRS
+        input_items = sorted(
+            [x, *y] for x, y in NON_CORE_ATTRS.items() if x not in set(mandatory_attrs)
         )
+        self.table_writer = TableWriter(
+            filename="generated/topology/topologyattrs.txt",
+            lines=[],
+            input_items=input_items,
+            column_spec=[
+                ("Atom", _atom),
+                ("AtomGroup", _atomgroup),
+                ("Description", _description),
+                ("Supported formats", _supported_formats),
+            ],
+        )
+        self.table_writer.generate_lines_and_write_table()
 
-    def _atom(self, name, singular, *args):
-        return singular
-
-    def _atomgroup(self, name, *args):
-        return name
 
-    def _description(self, name, singular, description):
-        return description
+class ConnectivityAttrs:
+    def __init__(self, attrs: dict[str, Any]) -> None:
+        def _atom(name: str) -> str:
+            return name
 
-    def _supported_formats(self, name, singular, description):
-        return ", ".join(sorted(self.attrs[name]))
+        def _atomgroup(name: str) -> str:
+            return name
 
+        def _supported_formats(name: str) -> str:
+            return ", ".join(sorted(attrs[name]))
 
-class ConnectivityAttrs(TopologyAttrs):
-    headings = ("Atom", "AtomGroup", "Supported formats")
-    filename = "generated/topology/connectivityattrs.txt"
+        input_items = [("bonds",), ("angles",), ("dihedrals",), ("impropers",)]
 
-    def _set_up_input(self):
-        inp = [[x] * 3 for x in "bonds angles dihedrals impropers".split()]
-        return inp
+        self.table_writer = TableWriter(
+            filename="generated/topology/connectivityattrs.txt",
+            input_items=input_items,
+            lines=[],
+            column_spec=[
+                ("Atom", _atom),
+                ("AtomGroup", _atomgroup),
+                ("Supported formats", _supported_formats),
+            ],
+        )
+        self.table_writer.generate_lines_and_write_table()
 
 
-if __name__ == "__main__":
+def main() -> None:
     top = TopologyParsers()
     TopologyAttrs(top.attrs)
     ConnectivityAttrs(top.attrs)
```

Comment:
In relation to suggested `attrs` change above, I don't think `main` would need to change a lot:

```suggestion
    top = TopologyParsers()
    topology_attrs = _get_format_attrs(top)
    TopologyAttrs(topology_attrs)
    ConnectivityAttrs(topology_attrs)
```